File: sdks/go/resource.go
package paymentflow

import (
	"context"
	"encoding/json"
	"fmt"
)

// The plumbing every resource service shares. Services are thin on purpose: a method names an
// operation, hands over its parameters, and returns what the API returned. Everything that
// could differ between them — how a path is filled in, which query parameters are legal, when
// an idempotency key is generated, what gets retried — lives in the transport.

func decodeError(o operationDescriptor, err error) error {
	return &APIError{&errorImpl{Message: fmt.Sprintf("could not decode the %s response: %s", o.ID, err)}}
}

// send issues one call and decodes the JSON object body into *T.
func send[T any](ctx context.Context, t *transport, o operationDescriptor, pathParams map[string]string,
	query map[string]any, body any, opts []RequestOption) (*T, error) {
	res, err := t.do(ctx, o, pathParams, query, body, newRequestConfig(opts))
	if err != nil {
		return nil, err
	}
	out := new(T)
	if len(res.body) > 0 {
		if err := json.Unmarshal(res.body, out); err != nil {
			return nil, decodeError(o, err)
		}
	}
	return out, nil
}

// sendList decodes a bare JSON array response element-wise.
func sendList[T any](ctx context.Context, t *transport, o operationDescriptor, pathParams map[string]string,
	opts []RequestOption) ([]T, error) {
	res, err := t.do(ctx, o, pathParams, nil, nil, newRequestConfig(opts))
	if err != nil {
		return nil, err
	}
	var out []T
	if len(res.body) > 0 {
		if err := json.Unmarshal(res.body, &out); err != nil {
			return nil, decodeError(o, err)
		}
	}
	return out, nil
}

// sendNoContent issues a call whose response body is discarded (a 204).
func sendNoContent(ctx context.Context, t *transport, o operationDescriptor, pathParams map[string]string,
	opts []RequestOption) error {
	_, err := t.do(ctx, o, pathParams, nil, nil, newRequestConfig(opts))
	return err
}

// q is a small builder for a query map that drops zero values.
type q map[string]any

func (m q) str(key, value string) q {
	if value != "" {
		m[key] = value
	}
	return m
}

func (m q) int64p(key string, value *int64) q {
	if value != nil {
		m[key] = *value
	}
	return m
}

func (m q) intp(key string, value *int) q {
	if value != nil {
		m[key] = *value
	}
	return m
}

func (m q) strs(key string, value []string) q {
	if len(value) > 0 {
		m[key] = value
	}
	return m
}

func (m q) meta(key string, value map[string]string) q {
	if len(value) > 0 {
		m[key] = value
	}
	return m
}

// Int64 returns a pointer to v — a convenience for the optional *int64 filter fields.
func Int64(v int64) *int64 { return &v }

// Int returns a pointer to v.
func Int(v int) *int { return &v }

File: sdks/java/src/main/java/dev/paymentflow/resources/Resource.java
package dev.paymentflow.resources;

import dev.paymentflow.CursorPage;
import dev.paymentflow.OffsetPage;
import dev.paymentflow.RequestOptions;
import dev.paymentflow.internal.Json;
import dev.paymentflow.internal.PageFetch;
import dev.paymentflow.internal.Paginator;
import dev.paymentflow.internal.RequestSpec;
import dev.paymentflow.internal.Transport;
import dev.paymentflow.model.Operations.OperationDescriptor;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

/**
 * What every resource namespace shares. Resource classes are thin on purpose: a method names an
 * operation, hands over its parameters, and returns what the API returned. Everything that could
 * differ between them — how a path is filled in, which query parameters are legal, when an
 * idempotency key is generated, what gets retried — lives in the transport, so adding an
 * endpoint cannot accidentally add a behaviour.
 */
public abstract class Resource {

    protected final Transport transport;

    protected Resource(Transport transport) {
        this.transport = transport;
    }

    /** One call, deserialized to {@code type}. */
    protected <T> T send(OperationDescriptor op, Map<String, String> path, Map<String, Object> query,
                         Object body, RequestOptions options, Class<T> type) {
        RequestSpec spec = new RequestSpec(op, path == null ? Map.of() : path,
                query == null ? Map.of() : query, body, options);
        Transport.Result result = transport.request(spec);
        return Json.toRecord(result.data(), type);
    }

    /** One call whose response is a bare JSON array, deserialized element-wise to {@code element}. */
    protected <T> List<T> sendList(OperationDescriptor op, Map<String, String> path, Map<String, Object> query,
                                   RequestOptions options, Class<T> element) {
        RequestSpec spec = new RequestSpec(op, path == null ? Map.of() : path,
                query == null ? Map.of() : query, null, options);
        Object data = transport.request(spec).data();
        if (!(data instanceof List<?> list)) {
            return List.of();
        }
        List<T> out = new ArrayList<>();
        for (Object raw : list) {
            out.add(Json.toRecord(raw, element));
        }
        return List.copyOf(out);
    }

    /** One call whose response body is discarded (a 204, or a fire-and-forget). */
    protected void sendVoid(OperationDescriptor op, Map<String, String> path, Object body, RequestOptions options) {
        transport.request(new RequestSpec(op, path == null ? Map.of() : path, Map.of(), body, options));
    }

    /** A cursor-paginated list. The filters are captured so every later page uses the same ones. */
    protected <T> CursorPage<T> listCursor(OperationDescriptor op, Map<String, Object> query,
                                           RequestOptions options, Class<T> element) {
        Function<Object, T> mapper = raw -> Json.toRecord(raw, element);
        PageFetch fetch = locator -> {
            Map<String, Object> page = new LinkedHashMap<>(query);
            if (locator != null) {
                page.put("starting_after", locator);
            }
            return transport.request(new RequestSpec(op, Map.of(), page, null, options));
        };
        return Paginator.cursor(fetch.fetch(null), fetch, mapper);
    }

    /** An offset-paginated list — the two endpoints D139 left on the older envelope. */
    protected <T> OffsetPage<T> listOffset(OperationDescriptor op, Map<String, String> path,
                                           Map<String, Object> query, RequestOptions options, Class<T> element) {
        Function<Object, T> mapper = raw -> Json.toRecord(raw, element);
        Map<String, String> pathParams = path == null ? Map.of() : path;
        PageFetch fetch = locator -> {
            Map<String, Object> page = new LinkedHashMap<>(query);
            page.put("page", locator == null ? 0 : locator);
            return transport.request(new RequestSpec(op, pathParams, page, null, options));
        };
        return Paginator.offset(fetch.fetch(null), fetch, mapper);
    }

    // ── small builders, so the resource methods read as data ────────────────────────────────

    protected static RequestOptions opts(RequestOptions options) {
        return options == null ? RequestOptions.NONE : options;
    }

    /** A query map that drops null values, in insertion order. */
    protected static QueryBuilder query() {
        return new QueryBuilder();
    }

    protected static final class QueryBuilder {

        private final Map<String, Object> map = new LinkedHashMap<>();

        public QueryBuilder put(String key, Object value) {
            if (value != null) {
                map.put(key, value);
            }
            return this;
        }

        public Map<String, Object> build() {
            return map;
        }
    }

    /** A body map that drops null values — the same rule as {@code body_of} in the Python SDK. */
    protected static BodyBuilder body() {
        return new BodyBuilder();
    }

    protected static final class BodyBuilder {

        private final Map<String, Object> map = new LinkedHashMap<>();

        public BodyBuilder put(String key, Object value) {
            if (value != null) {
                map.put(key, value);
            }
            return this;
        }

        public Map<String, Object> build() {
            return map;
        }
    }
}
